### ISSUE TITLE
Reconstructing Hi-C Embeddings

### ISSUE THREAD

**Issue by Jpickard1** [Noble-Lab/HiCFoundation#3]
My goal is to take one `.hic` file & the pretrained HiCFoundation weights and generate both the embedding and reconstruction of the full `.hic` file, which I understand should be the output of the decoder. Could you please advise me on how to best accomplish this or point me to codes you have for this task?

I see in the example from the HiCFoundation repository, the Inference API allows the user to extract embeddings from the decoder, but to me it does not appear as though the shape of the decoder outputs are matching the shape of the input matrix at the highest level of the decoder. I included a screenshot below, which was generated using the pretrained model with the default settings of the `API for embedding` section of the README.md file.

Here is my current code I am running and the outputs:
```
import torch
from inference.load_model import load_model,to_cuda,to_float,format_input

#configure input
model_path="/nfs/turbo/umms-indikar/shared/projects/foundation_models/hicfoundation/hicfoundation_pretrain.pth.tar" # specify the path of pre-trained model on your directory
input_row_size=1000 # specify the input matrix row size, should be a multiply of 16 (patch_size of HiCFoundation)
input_col_size=128 # specify the input matrix column size, should be a multiply of 16 (patch_size of HiCFoundation)
total_count = 100000000 # the total read of your HiC matrix. 
embed_depth=0 # Specified the embedding to use for your purpose, default: 0 (encoder output embeddings). 
#You can also specify ``k`` from 1 to 8 to indicate the output of k-th layer of decoder.

#total_count = None, If not sure, set to None
max_rand= total_count if total_count is not None else 100000000 
input_mat = torch.randint(0, max_rand, (input_row_size,input_col_size)) # specify the input size, random integer array with value randomly sampled from 0 to max_rand, with the size of (input_row_size,input_col_size).
input_mat = format_input(input_mat)

#load model
model = load_model(model_path,input_row_size,input_col_size)

#load to cuda, if you do not have gpu, you can comment the following line
input_mat = to_cuda(input_mat)
total_count = to_cuda(total_count)
model = to_cuda(model)

#convert to float to do computation
input_mat = to_float(input_mat)
total_count = to_float(total_count)
model = to_float(model)

#inference of HiCFoundation
input_mat = input_mat.unsqueeze(0) #add batch dimension

total_count = total_count.unsqueeze(0) if total_count is not None else None #add batch dimension

output_full = model(input_mat,total_count)
print(f"{input_mat.shape=}")
for level in range(9):
    print(f"{output_full[level].shape=}")
```

Printed output:
```
Position interpolate from 14x14 to 62x8
Loading pre-train encoder message!
Position interpolate for decoder from 14x14 to 62x8
Loading pre-train model decoder!

input_mat.shape=torch.Size([1, 3, 1000, 128])
output_full[level].shape=torch.Size([1, 62, 8, 1024])
output_full[level].shape=torch.Size([1, 62, 8, 512])
output_full[level].shape=torch.Size([1, 62, 8, 512])
output_full[level].shape=torch.Size([1, 62, 8, 512])
output_full[level].shape=torch.Size([1, 62, 8, 512])
output_full[level].shape=torch.Size([1, 62, 8, 512])
output_full[level].shape=torch.Size([1, 62, 8, 512])
output_full[level].shape=torch.Size([1, 62, 8, 512])
output_full[level].shape=torch.Size([1, 62, 8, 512])
```

I would expect the output from the decoder to be a similar shape or at least have the dimensions `(1000, 128)`, but instead the `(62, 8)` is approximately 16th the dimensions. I think this is likely due to the patch size being 16 (but am not certain). If this is the case, that would make sense to me because we have one 512 embedding vector per patch. I am curious how I can run perhaps another layer of the decoder to convert these embeddings into the patched / pixels that would be the size of the image as I expect.

Please let me know if I am misunderstanding a major portion of this code or how to proceed. Thank you for your assistance!

**Comment by wang3702** [comment]
Thank you for your interest in our work! 
I think if you revise the https://github.com/Noble-Lab/HiCFoundation/blob/5e7306b17cab23fb78d4e262254856d454dc70a4/inference/load_model.py#L23 task=2, and still use the pre-trained model for embeddings. You should be able to generate the reconstruction you want to have.

**Comment by Jpickard1** [comment]
@wang3702 thank you for your prompt reply!

That fix does give exactly what I need! I knew it would be a quick fix, but was struggling to get it myself. I made a PR [here](https://github.com/Noble-Lab/HiCFoundation/pull/4) #4 that adds an optional argument for the task in the file you referenced, but it keeps the default at `task=6` so as to not change your codes. Please let me know if this is an appropriate change.

I may follow up with some questions in the future, but I can make new issues for them if you prefer. For instance, do you know if it is possible to run the decoder of your model without first running in encoder? This is something I am looking to do, and curious if you have explored it yet.

**Comment by wang3702** [comment]
Thank you for your PR! I have merged it. 

The decoder needs to take input from encoder, but of course you can directly run the decoder with the input you wanted to have. But it may not makes too much sense. 
You can consider skip  https://github.com/Noble-Lab/HiCFoundation/blob/9622c002de2d9fb7f39cc647394e4053aac0f201/model/Finetune_Model_Head.py#L248 this function to only forward decoder if you want

**Comment by Jpickard1** [comment]
Dear @wang3702 thank you for the help setting up the reconstruction according to the **API for embedding** codes from the README.md. I am having trouble getting the reconstructions to look good and qualitatively resemble the input Hi-C matrices.

Currently, I have modified the above code to replace `input_mat` being random matrix to make `input_mat` a matrix loaded from a Hi-C matrix, but the reconstructions are not looking correct:

![Image](https://github.com/user-attachments/assets/4e519a0f-4ae0-4bd8-8bb2-8ac8f353d69f)


I am sure that I am making a major error somewhere, so I am wondering if you could share with me the codes you used to generate "Extended Figure S1" from your [paper](https://www.biorxiv.org/content/10.1101/2024.12.16.628821v1.full.pdf) so that I can see an example of how to do this or where my error is.

Please let me know if you can share codes to recreate these reconstructions in that figure. Thank you. Best, Joshua

**Comment by wang3702** [comment]
For the visualization in paper, I did not use this codebase. But you should get similar results with this codebase.
Could you please share your code details here so that I can take a look?
First, I believe there may be an issue with your output color scale. You can use the following code to convert any count data into RGB images: [GitHub link](https://github.com/Noble-Lab/HiCFoundation/blob/9622c002de2d9fb7f39cc647394e4053aac0f201/data_processing/inference_dataset.py#L142). Your input and output images should at least resemble a red-style colormap.

**Comment by Jpickard1** [comment]
One issue I notice off the bat is that you linked in your response to the `Inference_Dataset` class as being useful to make the color map red; however, this class is not used in the **API for embedding** section of the README, which I have been following to do these reconstructions. I am not sure how to be using this class based on the example code, but I put the code I used to generate these reconstructions according to the API for Embedding section below:

I have 4 blocks of code below:
1. Loading the Hi-C data
2. Loading the model
3. Running inference on the model
4. Visualizing the input and output side by side

## My Codes

**Loading the Hi-C data**
```
# Import standard packages
import os
import pickle
import torch
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm  # Import LogNorm

# Load Hi-C file that is processed using the `hic2array.py` function from https://github.com/Noble-Lab/HiCFoundation_paper/blob/main/utils/hic2array.py
file_path = "/nfs/turbo/umms-indikar/shared/projects/twin_cell/data/hicfoundation_v2/hic2array/ENCFF528JJC.pkl"
with open(file_path, 'rb') as file:
    data = pickle.load(file)

# Convert to 100kb matrix and get section to visualize
chr1 = data['chr1'].toarray()
chr1_100kb = chr1[:49700,:49700].reshape(497, 100, 497, 100).sum(axis=(1, 3))
lchr1 = np.log1p(chr1_100kb)
lchr1 += lchr1.T
vis = lchr1[:96,:96]

# Extract the piece of the matrix that we will generate reconstructions for
data = chr1_100kb[:96, :96]
input_mat = torch.tensor(data)

# Visualize the data we will pass through the model
plt.imshow(vis, cmap='Reds')
plt.colorbar()
plt.show()
```
**Loading the model**
The below code is taken almost directly from the README example, with the one update #4 to set `task=2` when loading the model:
```
from inference.load_model import load_model,to_cuda,to_float,format_input

#configure input
model_path="/nfs/turbo/umms-indikar/shared/projects/foundation_models/hicfoundation/hicfoundation_pretrain.pth.tar" # specify the path of pre-trained model on your directory
input_row_size=input_mat.shape[1] # specify the input matrix row size, should be a multiply of 16 (patch_size of HiCFoundation)
input_col_size=input_mat.shape[1] # specify the input matrix column size, should be a multiply of 16 (patch_size of HiCFoundation)
total_count = 100000000 # the total read of your HiC matrix. 
embed_depth=0 # Specified the embedding to use for your purpose, default: 0 (encoder output embeddings). 

# This line is different from the README, setting the task = 2
model = load_model(model_path,input_row_size,input_col_size,task=2)
input_mat = format_input(input_mat)

#load to cuda, if you do not have gpu, you can comment the following line
input_mat = to_cuda(input_mat)
total_count = to_cuda(total_count)
model = to_cuda(model)

#convert to float to do computation
input_mat = to_float(input_mat)
total_count = to_float(total_count)
model = to_float(model)

#inference of HiCFoundation
input_mat = input_mat.unsqueeze(0) #add batch dimension

total_count = total_count.unsqueeze(0) if total_count is not None else None #add batch dimension
```

**Running inference on the model**
```
output = model(input_mat,total_count)
```

**Visualization**
This code visualizes the input matrix data and the reconstruction. Note, the color scales are not the same, but qualitatively the shape of the matrix and the dominante features appear far off. The output of this is shown below.
```
reconstruction = output[0,:,:].cpu().detach().numpy()

fig, axes = plt.subplots(1, 2, figsize=(10, 5))

axes[0].imshow(np.log1p(data), cmap='Reds')  # Input on the left
axes[0].set_title("Input")
axes[0].axis("off")  # Hide axes for better visualization

axes[1].imshow(np.log1p(reconstruction), cmap='Reds')  # Output on the right
axes[1].set_title("Output")
axes[1].axis("off")

plt.tight_layout()
plt.show()
```

![Image](https://github.com/user-attachments/assets/ae521553-38f9-45a3-8b20-bc99b0ed7043)

## Possible issues:
I am wondering if any of these items seem out of place:
1. Perhaps I am inputing the Hi-C data into the model to run inference incorrectly. What I am doing is based on the README example, which is only using a random matrix 
2. Perhaps the model weights I am using are not correct
3. Perhaps there is an issue in data normalization / `.hic` preprocessing

Please let me know if you can reproduce this issue or what you think might be causing this. If you have an example where you do get good reconstructions from the pretrained weights, I would be interested to compare my code with that, such as the code you mentioned for the figure S1.

My apologies if this is tool long (I thought it best to share all the details), and I sincerely appreciate your assistance. I am extremely excited to get this working :)

**Comment by wang3702** [comment]
First, I think your input is wrong. You should input full matrix for the input instead of a upper diagonal input.
Second, my plot is at 5kb, I think you can first try that to see if you can reproduce. Though I feel 100kb should also be fine, it would be good first try at 5kb. 
Third, I think you can directly convert any resolution pkl file by hic2array.py. I don't understand why you wanted to do the sum by yourself.
Lastly, I would recommend you to check the values in output array and visualize it in raw space instead of the log space.
I feel I can see some patterns for your right output figure, but it may be hiden because of log or color scale.
```python
def convert_rgb(data,max_value):
    """
    The convert_rgb function takes in a 2D array and converts it to a 3D RGB array.
    :param data: Specify the input 2D array
    :param max_value: Specify the maximum threshold of the input array for figures
    :return: A 3D RGB array of size (data.shape[0], data.shape[1], 3)
    """
    data_red = np.ones(data.shape)
    data = np.minimum(data,max_value)
    data = (max_value-data)/max_value
    data_rgb = np.stack([data_red,data,data],axis=0,dtype=np.float32)#transform only accept channel last case
    data_rgb = data_rgb.transpose(1,2,0)
    data_rgb = (data_rgb*255).astype(np.uint8)
    return data_rgb

```

**Comment by Jpickard1** [comment]
Dear @wang3702 thank you! This is extremely helpful.

> First, I think your input is wrong. You should input full matrix for the input instead of a upper diagonal input.

Thank you for pointing this out, it makes a major difference

> Second, my plot is at 5kb, I think you can first try that to see if you can reproduce. Though I feel 100kb should also be fine, it would be good first try at 5kb.
Third, I think you can directly convert any resolution pkl file by hic2array.py. I don't understand why you wanted to do the sum by yourself.

I have revised my code below to only use 5kb resolution. I am having an issue running inference on the model for the whole chromosome as I am not able to fit everything in memory, but regardless, at 5kb I and 1000 bins it is still becoming a larger piece matrix to run inference on.

**Load Data**
```
import os
import pickle
import torch
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm  # Import LogNorm

file_path = "/nfs/turbo/umms-indikar/shared/projects/twin_cell/data/hicfoundation_v2/hic2array/ENCFF528JJC.pkl"
with open(file_path, 'rb') as file:
    data = pickle.load(file)
chr1_100kb = data['chr1'].toarray()
data = chr1_100kb[200:1200,200:1200]
data += data.T
input_mat = torch.tensor(data)
```

**Load Model**
```
import sys
sys.path.append("HiCFoundation")

import torch
from inference.load_model import load_model,to_cuda,to_float,format_input

model_path="/nfs/turbo/umms-indikar/shared/projects/foundation_models/hicfoundation/hicfoundation_pretrain.pth.tar"
input_row_size=input_mat.shape[1]
input_col_size=input_mat.shape[1]
total_count = 100000000
embed_depth=0

model = load_model(model_path, input_row_size, input_col_size, task=2)

input_mat = format_input(input_mat)
input_mat = to_float(input_mat)
total_count = to_float(torch.tensor(total_count))
model = to_float(model)
input_mat = input_mat.unsqueeze(0)
total_count = total_count.unsqueeze(0) if total_count is not None else None #add batch dimension
```

**Inference**
```
output = model(input_mat,total_count)
```

**Visualization**
```
def convert_rgb(data,max_value):
    """
    The convert_rgb function takes in a 2D array and converts it to a 3D RGB array.
    :param data: Specify the input 2D array
    :param max_value: Specify the maximum threshold of the input array for figures
    :return: A 3D RGB array of size (data.shape[0], data.shape[1], 3)
    """
    data_red = np.ones(data.shape)
    data = np.minimum(data,max_value)
    data = (max_value-data)/max_value
    data_rgb = np.stack([data_red,data,data],axis=0,dtype=np.float32)#transform only accept channel last case
    data_rgb = data_rgb.transpose(1,2,0)
    data_rgb = (data_rgb*255).astype(np.uint8)
    return data_rgb

reconstruction = output[0,:,:].cpu().detach().numpy()
fig, axes = plt.subplots(1, 2, figsize=(10, 5))
axes[0].imshow(convert_rgb(data, 256))
axes[0].set_title("Input (1000 loci at 5kb)")
axes[1].imshow(convert_rgb(reconstruction, 256))
axes[1].set_title("Output")
plt.tight_layout()
plt.show()
```

This is the output of the raw data using your visualization code note in the log space:

![Image](https://github.com/user-attachments/assets/cb21ab6a-29d7-4fb9-b2ce-7d488c81430f)

The raw data visualized in the log space looks like this:

![Image](https://github.com/user-attachments/assets/31e04e16-ad12-4218-a209-587829beedd7)

I do agree that it seems the reconstruction is showing some of the structure from the raw data, although my visualizations do not look as nice as those in your paper yet. These still seems to be major features and structure missing from the reconstruction.

It seems including a larger matrix of Hi-C when running inference does improve structure of the reconstruction, but I am having an issue putting a full chromosome through the pretrained model at 5kb resolution. Do you have any advice on how to keep the data at 5kb and use a full chromosome without running into memory constraints on the GPU or CPU? Do you know how much RAM you needed to perform this?

Separately, I know you mentioned the code you used for extended figure 1 are not shared in this repo or the paper repo, but I'm wondering if you are able to share these codes with me, at least off line or via a meeting? It would be extremely helpful for a project I am working on if we could make reconstructions as nice as those shown in your figures.

Please let me know your next suggestions. Thank you for your time! Best, Joshua

**Comment by wang3702** [comment]
My bad, task=2 is not so suitable for reconstruction. I have added a task=7 for your purpose. The testing code is attached:
```python3
import os
import pickle
import torch
import numpy as np
import matplotlib.pyplot as plt

file_path="tmp_test/ENCFF528JJC.pkl"
data=pickle.load(open(file_path,"rb")) 
chr1=data['chr1_chr1'].toarray()
input_mat=chr1[10000:10224,10000:10224] 
input_mat=input_mat+np.triu(input_mat,1).T
input_vis = np.array(input_mat)
from inference.load_model import load_model,to_cuda,to_float,format_input
model_path="hicfoundation_model/hicfoundation_pretrain.pth.tar"

input_row_size=input_mat.shape[1]
input_col_size=input_mat.shape[1]
total_count = 100000000
embed_depth=0

model = load_model(model_path, input_row_size, input_col_size, task=7)
input_mat = torch.tensor(input_mat)
input_mat = format_input(input_mat)
input_mat = to_float(input_mat)
total_count = to_float(torch.tensor(total_count))
model = to_float(model)
input_mat = input_mat.unsqueeze(0)
total_count = total_count.unsqueeze(0) if total_count is not None else None #add batch dimension

output = model(input_mat,total_count)

def convert_rgb(data,max_value):
	"""
	The convert_rgb function takes in a 2D array and converts it to a 3D RGB array.
	:param data: Specify the input 2D array
	:param max_value: Specify the maximum threshold of the input array for figures
	:return: A 3D RGB array of size (data.shape[0], data.shape[1], 3)
	"""
	data_red = np.ones(data.shape)
	data = np.minimum(data,max_value)
	data = (max_value-data)/max_value
	data_rgb = np.stack([data_red,data,data],axis=0,dtype=np.float32)#transform only accept channel last case
	data_rgb = data_rgb.transpose(1,2,0)
	data_rgb = (data_rgb*255).astype(np.uint8)
	return data_rgb
print(input_vis)
print(output)
reconstruction = output[0,:,:].cpu().detach().numpy()
reconstruction = reconstruction.astype(np.uint8)
reconstruction = reconstruction.transpose(1,2,0)
fig, axes = plt.subplots(1, 2, figsize=(10, 5))
axes[0].imshow(convert_rgb(input_vis, 20))
axes[0].set_title("Input (1000 loci at 5kb)")
axes[1].imshow(reconstruction)
axes[1].set_title("Output")
plt.tight_layout()
plt.savefig("tmp.png",dpi=600)
```
The visualization is attached for your reference.

![Image](https://github.com/user-attachments/assets/261a458c-0eea-477a-8788-132314b5d13a)
